lib/mix/lib/mix/rebar.ex

```diff
@@ -38,6 +38,26 @@ defmodule Mix.Rebar do
     global_rebar_cmd(manager) || local_rebar_cmd(manager)
   end
 
+  @doc """
+  Returns the version of the available `rebar` command if any
+  """
+  @rebar_vsn_regex ~r/^## Rebar3 (?<vsn>[\w\.\d]+)$/m
+  def rebar_vsn(path_to_rebar) do
+    maybe_vsn_line =
+      path_to_rebar
+      |> File.stream!()
+      |> Stream.take(2)
+      |> Enum.at(1)
```

Comment:
Is this equivalent to:

```suggestion
      |> Enum.at(3)
```

?

In any case, I am worried this will be too brittle. I wonder if we can ask Rebar to support an environment variable instead? The environment variable is preferred because we can pass it to both old and new rebar versions alike, then we just check to write directory it wrote to.

Thoughts?

Reply:
Before I forget, thank you for looking into this, it is really appreciated and it will definitely improve the UX.

Reply:
It is equivalent to:

```suggestion
      |> Enum.at(1)
```

The `Stream.take` call is redundant.

Reply:
Does rebar error on unknown command line arguments? If it doesn't then we don't have to use environment variables and can just check the directory.

Reply:
It does.

Reply:
How would rebar3 support an environment variable there? I'm not sure what is meant by that.

Reply:
Also the annotation, if present, is always on the second line of the file. asdf already annotated on that line, and we added one for both the escript and the local install.

Reply:
@ferd something like REBAR3_OUTPUT_DIR, if the flag is not set, you would look for this env var. :)

Reply:
Nothing would stop us from supporting that, but it wouldn't really fit most of the configuration values we have (we tend to use globals for values that impact core functionality outside of specific commands and have no specific switches). The OUTPUT_DIR naming is also tricky since it has an ambiguous naming in contexts other than the bare compiler (we'd need REBAR3_BARE_COMPILER_OUTPUT_DIR maybe), and requires explicit documenting outside of just the --help switches.

Overall I feel like it's a pretty bad fit to use the env variables for that from our end of things. That being said it's not like the current annotation isn't a big hack, but it's at least pretty invisible and more general for version/feature detection for future needs people could have.

Reply:
A very long variable name would be fine.

I am just a bit worried about reading the version from parsing the executable - it feels a bit fragile. We could execute `rebar version` but that adds to compilation times. So without a more "transparent" mechanism, it may be hard for us to support different rebar versions.

Reply:
It shouldn't be that fragile and if it fails it just falls back to compiling the same as it has been doing, so nothing gets worse if version isn't found.

Of course other option is to fall back on running `rebar3 version` if the parsing fails, just in case, and then falling back to compiling as it does today.

Reply:
I'm ok with either way. One thing that will have to be done per failing CI is detect if are win32 or not, and if so split the rebar_path because the rebar path for win32 includes `"escript.exe \C:\..."`

Reply:
Hmm yeah, so I think while getting the rebar version from the file is probably the best way to go. The only thing I can think of to improve that is to get the version when rebar3 is retrieved via mix and stick it in `.mix` dir as `REBAR_VERSION`